AnastasiaB/cards4-template.js

```diff
@@ -0,0 +1,79 @@
+
+
+var makeCard = // receive factory with external name `makeCard`
+    (function () { //begin IIFE...
+
+    // The factory itself:
+ function makeCard(id) {
+  if (!makeCard.isValidID(id))
+   return null;
+
+    var cards = {
+      id: id,
+      rank: rank,
+      suit: suit,
+      color: color,
+      cardName: cardName
+    };
+
+    return cards;
+    }
+
+//--------------------------
+// Private resources (internal use only)
+//--------------------------
+
+    // Examples:
+
+makeCard.isValidID = function(num) {
+ return ((typeof num)==="number") && (num%1 === 0) && num>=0 && num<=51;
+};
+
+ var rankNames = ['','Ace','Two','Three','Four','Five','Six','Seven','Eight','Nine','Ten','Jack','Queen','King'];
+ var suitNames = ['','Hearts','Diamonds','Spades','Clubs'];
+
+//-----------------------
+// Instance Methods:
+//-----------------------
+
+    function rank() {
+      return Math.floor(this.id / 4) + 1;
+    }
+
+    function suit() {
+      return ((this.id % 4) + 1);
+    }
+
+    function color() {
+     var theSuit = this.suit();
+      return theSuit && ((theSuit < 3) ? "red" : "black");
+    }
+
+    function cardName() {
+     var theRank = this.rank();
+     var theSuit = this.suit();
+     return rankNames[theRank] + ' of ' + suitNames[theSuit]; //--> string
+
+}
+
+//-----------------------
+// Factory Methods/Data:
+//-----------------------
+
+    makeCard.isCard = function(thing) {
+     return (typeof thing === 'object') && (makeCard.rank === thing.rank) && ('id' in thing) && (makeCard.isValidID(thing.id));
+    };
+
+    makeCard.fullSet = [];//<-- fill me
+     for (var id=0; id<=52; id++) {
```

Comment:
In its current state makeCard.fullSet is going to give you an array with 1 null object at the end. The reason is that you're starting at id 0 and ending at id 52 (while including 52 as well), but remember that this is actually 53 ids, which in turn is going to be rejected by your isValid function because it exceeds your num limit (of 51).